### PR TITLE
Interface to MKL Parallel Direct Sparse Solver for Clusters 

### PR DESCRIPTION
**Pull request by jandrej**
- [x] Make build system
- [x] Changelog

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1784](https://github.com/mfem/mfem/pull/1784) | @jandrej | @tzanio | @jamiebramwell + @psocratis | 09/27/20 | 10/04/20 | 10/12/20 | |

### CODE REVIEW

File: linalg/cpardiso.hpp [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    * @param mat_type Matrix type
+    */
+   void SetMatrixType(MatType mat_type);
+
+   ~CPardisoSolver();
+
+private:
+   MPI_Comm comm_;
+
+   // Global numer of rows
```
Comment (jandrej) [inline linalg/cpardiso.hpp:84]: typo

File: examples/makefile [hunk 2]
```diff
@@ -119,6 +119,11 @@ ex11p-test-superlu: ex11p
 	@$(call mfem-test,$<, $(RUN_MPI), SuperLU_DIST example,--superlu)
 test-par-YES: ex11p-test-superlu
 endif
+ifeq ($(MFEM_USE_MKL_CPARDISO),YES)
+ex11p-test-superlu: ex11p
```
Comment (jamiebramwell) [inline examples/makefile:123]: `ex11p-test-pardiso`

File: linalg/cpardiso.hpp [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+
+   // Dummy variables
+   mutable int idum;
+   mutable double ddum;
+};
+} // namespace mfem
+
+#endif
+#endif // MFEM_USE_MKL_CPARDISO
+#endif // MFEM_USE_MPI
```
Comment (jamiebramwell) [inline linalg/cpardiso.hpp:124]: newline

File: linalg/cpardiso.cpp [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+   delete[] csr_rowptr;
+   delete[] reordered_csr_colind;
+   delete[] reordered_csr_nzval;
+}
+
+} // namespace mfem
+
+#endif // MFEM_USE_MKL_CPARDISO
+#endif // MFEM_USE_MPI
```
Comment (jamiebramwell) [inline linalg/cpardiso.cpp:230]: newline

**Review by jamiebramwell — COMMENTED** [review]
This looks great! I can't wait to give it a try! I just had a few typo comments.

**Review by psocratis — APPROVED** [review]
Looks very good!

**Review by jamiebramwell — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: INSTALL [hunk 5]
```diff
@@ -486,6 +486,13 @@ MFEM_USE_CEED = YES/NO
    library for performant high-order operator evaluation developed by the Center
    for Efficient Exascale Discretizations in the Exascale Computing Project.
 
+MFEM_USE_MKL_CPARDISO = YES/NO
+   Enables the interface to the Intel MKL Parallel Direct Sparse Solver for
+   Clusters. Make sure to set the correct values for MKL_MPI_WRAPPER and
+   MKL_LIBRARY_SUBDIR as shown in defaults.mk. If you configure MFEM with
+   MFEM_USE_LAPACK=YES, verify that the MKL LAPACK libraries are used. The
+   OpenMP capabilities are disabled at link time.
+
```
Comment (tzanio) [inline INSTALL:495]: Can you clarify the CPARDISO portion of the name, e.g. "the Intel MKL Parallel Direct Sparse Solver for Clusters (CPardiso)"?

What's the relation between the MKL CPARDISO and just PARDISO (https://www.pardiso-project.org/)?

File: INSTALL [hunk 6]
```diff
@@ -486,6 +486,13 @@ MFEM_USE_CEED = YES/NO
    library for performant high-order operator evaluation developed by the Center
    for Efficient Exascale Discretizations in the Exascale Computing Project.
 
+MFEM_USE_MKL_CPARDISO = YES/NO
```
Comment (tzanio) [inline INSTALL:489]: Need to add also a section under "External Libraries" with URL, version requirements, etc.

File: CHANGELOG [hunk 7]
```diff
@@ -137,6 +137,9 @@ Linear and nonlinear solvers
 - Added partially assembled convergent diagonal preconditioner for adaptively
   refined meshes (i.e. non-conforming finite element spaces), see Example 6/6p.
 
+- Added an interface to the Intel MKL Parallel Direct Sparse Solver for
+  Clusters. An example usage of the interface is shown in Example 11p.
```
Comment (tzanio) [inline CHANGELOG:141]: Maybe clarify that this is a parallel direct solver?

I'm not sure how many people will be familiar with this name... In fact calling it MKL CPardiso than the full name maybe more clear.

File: examples/ex11p.cpp [hunk 8]
```diff
@@ -95,6 +96,14 @@ int main(int argc, char *argv[])
...
    args.AddOption(&sp_solver, "-sp", "--strumpack", "-no-sp",
                   "--no-strumpack", "Use the STRUMPACK Solver.");
+#endif
+#ifdef MFEM_USE_MKL_CPARDISO
+   args.AddOption(&pardiso_solver,
+                  "-pardiso",
+                  "--pardiso",
+                  "-no-pardiso",
+                  "--no-pardiso",
+                  "Use the MKL Cluster Pardiso Solver.");
```
Comment (tzanio) [inline examples/ex11p.cpp:106]: Can we use a consistent style?

```suggestion
   args.AddOption(&pardiso_solver, "-pardiso", "--pardiso", "-no-pardiso",
                  "--no-pardiso", "Use the MKL Cluster Pardiso Solver.");
```

File: examples/ex11p.cpp [hunk 9]
```diff
@@ -268,10 +277,19 @@ int main(int argc, char *argv[])
...
       }
+#endif
+#ifdef MFEM_USE_MKL_CPARDISO
+      if (pardiso_solver)
+      {
+         auto pardiso = new CPardisoSolver(A->GetComm());
+         pardiso->SetMatrixType(CPardisoSolver::MatType::REAL_STRUCTURE_SYMMETRIC);
+         pardiso->SetPrintLevel(1);
+         pardiso->SetOperator(*A);
+         precond = pardiso;
```
Comment (tzanio) [inline examples/ex11p.cpp:288]: Nice and clean!

File: linalg/cpardiso.cpp [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+   iparm[39] = 2;
+   // Maximum number of numerical factorizations
+   maxfct = 1;
+   // Which factorization to use
+   mnum = 1;
+   // Print statistical information in file
+   msglvl = 0;
+   // Initialize error flag
+   error = 0;
+   // Real unsymmetric matrix
```
Comment (tzanio) [inline linalg/cpardiso.cpp:46]: Strange name, most people use "nonsymmetric"

File: linalg/cpardiso.cpp [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+   std::iota(permutation_idx.begin(), permutation_idx.end(), 0);
+   for (int i = 0; i < m_loc; i++)
+   {
+      std::sort(permutation_idx.begin() + csr_rowptr[i],
+                permutation_idx.begin() + csr_rowptr[i + 1],
+                [csr_colind](int i1, int i2)
+      {
+         return csr_colind[i1] < csr_colind[i2];
+      });
+   }
```
Comment (tzanio) [inline linalg/cpardiso.cpp:100]: 😆

File: linalg/cpardiso.cpp [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+                         &idum,
+                         &nrhs,
+                         iparm,
+                         &msglvl,
+                         &ddum,
+                         &ddum,
+                         &comm_,
+                         &error);
+
+   MFEM_ASSERT(error == 0, "Pardiso analyze input error");
```
Comment (tzanio) [inline linalg/cpardiso.cpp:138]: "Pardiso" or "CPardiso"? (There are a few other instances of this)

**Review by tzanio — COMMENTED** [review]
I have a few small editing suggestions below. Otherwise looks good!

File: INSTALL (hunk 5 above)
Comment (jandrej) [reply INSTALL:495]: MKL CPardiso and the parallel version of Pardiso are developed independently.

**Review by tzanio — APPROVED** [review]
Excellent addition, thank you @jandrej!

@jamiebramwell and @psocratis -- thank you for reviewing!

**Pull request merged**